src/python/grpcio_tests/tests_aio/unit/client_unary_stream_interceptor_test.py

```diff
@@ -0,0 +1,404 @@
+            test_constants.SHORT_TIMEOUT)
+
+    return validation()
+
+
+class _UnaryStreamInterceptorEmpty(aio.UnaryStreamClientInterceptor):
+
+    async def intercept_unary_stream(self, continuation, client_call_details,
+                                     request):
+        return await continuation(client_call_details, request)
+
+
+class _UnaryStreamInterceptorWith_ResponseIterator(
+        aio.UnaryStreamClientInterceptor):
+
+    def __init__(self):
+        self.response_iterator = None
+
+    async def intercept_unary_stream(self, continuation, client_call_details,
+                                     request):
+        call = await continuation(client_call_details, request)
+        self.response_iterator = _ResponseIterator(call)
+        return self.response_iterator
+
+
+class TestUnaryStreamClientInterceptor(AioTestBase):
+
+    async def setUp(self):
+        self._server_target, self._server = await start_test_server()
+
+    async def tearDown(self):
+        await self._server.stop(None)
+
+    async def test_intercepts(self):
+        for interceptor_class in (_UnaryStreamInterceptorEmpty,
+                                  _UnaryStreamInterceptorWith_ResponseIterator):
+
+            with self.subTest(name=interceptor_class):
+                interceptor = interceptor_class()
+
+                request = messages_pb2.StreamingOutputCallRequest()
+                for _ in range(_NUM_STREAM_RESPONSES):
+                    request.response_parameters.append(
+                        messages_pb2.ResponseParameters(
+                            size=_RESPONSE_PAYLOAD_SIZE))
+
+                channel = aio.insecure_channel(self._server_target,
+                                               interceptors=[interceptor])
+                stub = test_pb2_grpc.TestServiceStub(channel)
+                call = stub.StreamingOutputCall(request)
+
+                await call.wait_for_connection()
+
+                response_cnt = 0
+                async for response in call:
+                    response_cnt += 1
+                    self.assertIs(type(response),
+                                  messages_pb2.StreamingOutputCallResponse)
+                    self.assertEqual(_RESPONSE_PAYLOAD_SIZE,
+                                     len(response.payload.body))
+
+                self.assertTrue(response_cnt, _NUM_STREAM_RESPONSES)
+                self.assertEqual(await call.code(), grpc.StatusCode.OK)
+                self.assertEqual(await call.initial_metadata(), ())
+                self.assertEqual(await call.trailing_metadata(), ())
+                self.assertEqual(await call.details(), '')
+                self.assertEqual(await call.debug_error_string(), '')
+                self.assertEqual(call.cancel(), False)
+                self.assertEqual(call.cancelled(), False)
+                self.assertEqual(call.done(), True)
+
+                if interceptor_class == _UnaryStreamInterceptorWith_ResponseIterator:
```

Comment:
Nit: This check probably belongs in `_UnaryStreamInterceptorWith_ResponseIterator`. Maybe add an `assertInFinalState(self, test: unittest.TestCase)` method to your test interceptors? And then just call that?

Reply:
Ummm, how this would work? What would be the main benefit?

Reply:
This test case is parameterized on the `interceptor_class`, so ideally, it shouldn't have any specialized logic for each one because it violates information hiding.

It would basically be

```python
class _UnaryStreamInterceptorWith_ResponseIterator:
  ...
  def assert_in_final_state(self, test: unittest.TestCase):
    test.assertEqual(_NUM_STREAM_RESPONSES, interceptor.response_iterator.response_cnt)

...

class _UnaryStreamInterceptorEmpty:
  ...
  def assert_in_final_state(self, test):
    pass

...

def test_intercepts(self):
  with self.subtest(...):
    ...
    interceptor.assert_in_final_state(self)
```

I marked this one a nit because I don't think it's a blocker for merge, but if we continue adding special checks for each of the interceptor types, it kind of defeats the point of having a parameterized `subtest`.

Reply:
Thanks for sharing, I will update the test with your recommendation.

Reply:
done